wgpu/src/backend/wgpu_core.rs

```diff
@@ -2991,6 +2997,77 @@ fn default_error_handler(err: crate::Error) {
     panic!("wgpu error: {err}\n");
 }
 
+fn span_to_source_span(span: &naga::Span) -> Option<SourceSpan> {
+    span.to_range().map(|range| SourceSpan {
+        offset: range.start as u32,
+        length: range.len() as u32,
+    })
+}
+
+impl From<&CreateShaderModuleError> for CompilationInfo {
+    fn from(value: &CreateShaderModuleError) -> Self {
+        let message = match value {
+            #[cfg(feature = "wgsl")]
+            CreateShaderModuleError::Parsing(e) => CompilationMessage {
+                message: e.to_string(),
+                message_type: CompilationMessageType::Error,
+                span: e
+                    .inner
+                    .labels()
+                    .next()
+                    .and_then(|(span, _)| span_to_source_span(&span)),
+            },
+            #[cfg(feature = "glsl")]
+            CreateShaderModuleError::ParsingGlsl(e) => CompilationMessage {
+                message: e.to_string(),
+                message_type: CompilationMessageType::Error,
+                span: e
+                    .inner
+                    .errors
+                    .first()
+                    .and_then(|e| span_to_source_span(&e.meta)),
+            },
+            #[cfg(feature = "spirv")]
+            CreateShaderModuleError::ParsingSpirV(e) => CompilationMessage {
+                message: e.to_string(),
+                message_type: CompilationMessageType::Error,
+                span: None,
+            },
+            CreateShaderModuleError::Validation(e) => CompilationMessage {
+                message: e.to_string(),
+                message_type: CompilationMessageType::Error,
+                span: e
+                    .inner
+                    .spans()
+                    .next()
+                    .and_then(|(span, _)| span_to_source_span(span)),
+            },
+            CreateShaderModuleError::MissingFeatures(_)
+            | CreateShaderModuleError::InvalidGroupIndex { .. } => CompilationMessage {
+                message: value.to_string(),
+                message_type: CompilationMessageType::Error,
+                span: None,
+            },
+            // Device errors are reported through the error sink, and are not compilation errors.
+            // Same goes for native shader module generation errors.
+            CreateShaderModuleError::Device(_) | CreateShaderModuleError::Generation => {
```

Comment:
The WebGPU spec says
> A shader-creation error is an error feasibly detectable at shader module creation time. Detection relies only on the WGSL module source text and other information available to the createShaderModule API method. Statements in this specification that describe something the program must do generally produce a shader-creation error if those assertions are violated.

I figured that this is in the spirit of the specification, but am happy to change it.